### PR TITLE
Add GitHub webhook validation

### PR DESCRIPTION
**Pull request by enarha**
Implements https://docs.github.com/en/webhooks-and-events/webhooks/securing-your-webhooks Expects webhook secret to be made available through the `GH_APP_WEBHOOK_SECRET` environment variable. For local testing the validation functionality can be disabled though `insecure-skip-webhook-verify` flag.

### CODE REVIEW

**Comment by enarha** [comment]
/hold
This change should not be deployed by infra-deployments before https://github.com/redhat-appstudio/infra-deployments/pull/1721 is merged. It can be reviewed though.

**Comment by enarha** [comment]
/unhold
https://github.com/redhat-appstudio/infra-deployments/pull/1721 has been merged.

**Comment by adambkaplan** [comment]
@enarha please rebase your PR to address the merge conflicts.

File: pkg/proxy/proxy.go [hunk 1]
```diff
@@ -51,6 +53,14 @@ func (p *SprayProxy) HandleProxy(c *gin.Context) {
 		zap.String("query", c.Request.URL.RawQuery),
 		zap.Bool("insecure-tls", p.insecureTLS),
 		zap.String("request-id", c.GetString("requestId")),
+		zap.Bool("insecure-webhook", p.insecureWebhook),
+	}
+	// validate incoming request
+	if err := validateProxyRequest(c.Request, p.insecureWebhook); err != nil {
+		// we do not want to expose internal information, so returning generic failure message
+		c.String(http.StatusForbidden, "forbidden: request validation failure")
```
Comment (adambkaplan) [inline pkg/proxy/proxy.go]: I think returning a `400 Bad Request` code would be sufficient.

**Review by adambkaplan — COMMENTED** [review]
Generally looks good.

One request - when you rebase, please put the vendor and go.mod/sum changes in a separate commit. This makes it easier to manage future rebases and isolate code changes from dependency updates.

File: pkg/proxy/proxy.go (hunk 1 above)
Comment (enarha) [reply pkg/proxy/proxy.go]: Updated.

File: pkg/proxy/validate.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+func validateProxyRequest(req *http.Request, insecureWebhook bool) error {
+	if err := validateWebhookSignature(req, insecureWebhook); err != nil {
+		return err
+	}
+	return nil
+}
+
+func validateWebhookSignature(req *http.Request, insecureWebhook bool) error {
+	if !insecureWebhook {
+		if secret := os.Getenv(envWebhookSecret); secret != "" {
```
Comment (adambkaplan) [inline pkg/proxy/validate.go]: nit (non-blocking) - the secret should be passed into this function as a parameter. This makes code easier to unit test.

I consider best practice to read configuration options (like env vars) as close to the main entrypoint as possible, and then pass those down to the "business logic" through struct objects and function arguments.

**Review by adambkaplan — APPROVED** [review]
/approve

/lgtm

File: pkg/proxy/validate.go (hunk 2 above)
Comment (enarha) [reply pkg/proxy/validate.go]: That was my initial implementation, specifically with the unit testing in mind. Then I was thrown back by the increasing number of parameters passed around and specifically at NewServer and NewSprayProxy. Anyway, that makes sense and I'll refactor tomorrow.
Now I also pay attention that I diverged from what we agreed on the ticket, that the server should panic if the secret is not found and here it only affects the specific request. I'll fix that as well.

Comment (enarha) [reply pkg/proxy/validate.go]: I made the secret param of validateWebhookSignature.
Management of the secret env var was moved as "early" as the start of NewSprayProxy. The current implementation requires both insecureWebhook config and the actual secret and I did not want to pass 2 extra params from "main" -> NewServer -> NewSprayProxy.
Having the validation enabled (default) and missing the secret now makes the server to exit.

File: pkg/server/server_test.go [hunk 3]
```diff
@@ -9,22 +9,35 @@ import (
...
 	"net/http/httptest"
+	"net/url"
 	"strings"
 	"testing"
 
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
 
+// create GitHub webwook like HTTP request including signature
```
Comment (AndrienkoAleksandr) [inline pkg/server/server_test.go]:
```suggestion
// create GitHub webhook like HTTP request including signature
```

File: cmd/server.go [hunk 4]
```diff
@@ -55,6 +56,8 @@ sprayproxy server --backend http://localhost:8081 --backend http://localhost:808
 		metricsSrvr.StopServer()
 		return err
 	},
+	// prevent showing usage on RunE return
+	SilenceUsage: true,
```
Comment (adambkaplan) [inline cmd/server.go]: Why do we need this? I don't usually see this with other projects that use cobra.

Comment (enarha) [reply cmd/server.go]: When testing locally, existing on error (not cmd error) was printing the usage which felt odd. Other than that I did not find this affecting affecting cmd. I also found this issue, so it was not only me https://github.com/spf13/cobra/issues/340 . I do not have strong opinion and can revert if you prefer.

Comment (adambkaplan) [reply cmd/server.go]: Let's keep it, but update the comment so the justification is clear:

```suggestion
	// prevent showing usage if RunE returns an error - see https://github.com/spf13/cobra/issues/340
	SilenceUsage: true,
```

Comment (enarha) [reply cmd/server.go]: Updated.

**Review by AndrienkoAleksandr — APPROVED** [review]

**Comment by enarha** [comment]
@adambkaplan Any other comments/concerns on this PR?

File: pkg/proxy/proxy.go [hunk 5]
```diff
@@ -22,23 +23,40 @@ import (
 	"go.uber.org/zap/zapcore"
 )
 
-// GitHub webhook request max size is 25MB
-const maxReqSize = 1024 * 1024 * 25
+const (
+	// GitHub webhook request max size is 25MB
+	maxReqSize = 1024 * 1024 * 25
```
Comment (adambkaplan) [inline pkg/proxy/proxy.go:28]: Follow-up - we should make this configurable. I'm not going to block merge on this.

Created to https://issues.redhat.com/browse/PLNSRVCE-1312 to track.

**Review by adambkaplan — APPROVED** [review]
/lgtm

**Comment by adambkaplan** [comment]
Merging this PR as is - follow up item added to JIRA.

**Pull request merged**